File: cocos2d-x-2.2.0/projects/Ironman/Classes/CCShake.cpp
// Code by Francois Guibert  
// Contact: www.frozax.com - http://twitter.com/frozax - www.facebook.com/frozax  
  
  
#include "CCShake.h"  
#include "cocos2d.h"  
USING_NS_CC;  
  
// not really useful, but I like clean default constructors  
CCShake::CCShake() : m_strength_x(0), m_strength_y(0), m_initial_x(0), m_initial_y(0)  
{  
}  
  
CCShake* CCShake::create( float d, float strength )  
{  
    // call other construction method with twice the same strength  
    return createWithStrength( d, strength, strength );  
}  
  
CCShake* CCShake::createWithStrength(float duration, float strength_x, float strength_y)  
{  
    CCShake *pRet = new CCShake();  
      
    if (pRet && pRet->initWithDuration(duration, strength_x, strength_y))  
    {  
        pRet->autorelease();  
    }  
    else  
    {  
        CC_SAFE_DELETE(pRet);  
    }  
  
         
    return pRet;  
}  
  
bool CCShake::initWithDuration(float duration, float strength_x, float strength_y)  
{  
    if (CCActionInterval::initWithDuration(duration))  
    {  
        m_strength_x = strength_x;  
        m_strength_y = strength_y;  
        return true;  
    }  
      
    return false;  
}  
  
// Helper function. I included it here so that you can compile the whole file  
// it returns a random value between min and max included  
static float fgRangeRand( float min, float max )  
{  
    float rnd = ((float)rand()/(float)RAND_MAX);  
    return rnd*(max-min)+min;  
}  
  
void CCShake::update(float dt)  
{  
    float randx = fgRangeRand( -m_strength_x, m_strength_x )*dt;  
    float randy = fgRangeRand( -m_strength_y, m_strength_y )*dt;  
      
    // move the target to a shaked position  
    m_pTarget->setPosition( ccpAdd(ccp(m_initial_x, m_initial_y),ccp( randx, randy)));  
}  
  
void CCShake::startWithTarget(CCNode *pTarget)  
{  
    CCActionInterval::startWithTarget( pTarget );  
      
    // save the initial position  
    m_initial_x = pTarget->getPosition().x;  
    m_initial_y = pTarget->getPosition().y;  
}  
  
void CCShake::stop(void)  
{  
    // Action is done, reset clip position  
    this->getTarget()->setPosition( ccp( m_initial_x, m_initial_y ) );  
      
    CCActionInterval::stop();  
} 

File: cocos2d-html5_i/IronCity/src/other/CCShake.js
/**
 * Created with JetBrains WebStorm.
 * User: cocos
 * Date: 13-10-29
 * Time: 下午1:38
 * To change this template use File | Settings | File Templates.
 */

var CCShake = cc.ActionInterval.extend({
    m_strength_x:0,
    m_strength_y:0,
    m_initial_x:0,
    m_initial_y:0,
    m_pTarget:null,
    initWithDur:function(duration, strength_x, strength_y){
        this.m_strength_x = 0;
        this.m_strength_y = 0;
        this.m_initial_x = 0;
        this.m_initial_y = 0;

        if (this.initWithDuration(duration))
        {
            this.m_strength_x = strength_x;
            this.m_strength_y = strength_y;
            return true;
        }

        return false;
    },
    startWithTarget:function(pTarget){
        //this.startWithTarget( pTarget );
        this.m_pTarget = pTarget;

        // save the initial position
        this.m_initial_x = pTarget.getPositionX();
        this.m_initial_y = pTarget.getPositionY();
    },
    stop:function(){
        if(this.m_pTarget)
        {
            this.m_pTarget.stopAllActions();
            this.m_pTarget.setPosition( cc.p( this.m_initial_x, this.m_initial_y ) );
        }
    },
    update:function(dt){
        var randx = fgRangeRand( -this.m_strength_x, this.m_strength_x )*dt;
        var randy = fgRangeRand( -this.m_strength_y, this.m_strength_y )*dt;

        // move the target to a shaked position
        this.m_pTarget.setPosition( cc.pAdd(cc.p(this.m_initial_x, this.m_initial_y), cc.p( randx, randy)));
    }
});

var fgRangeRand  = function(min, max){
    var rnd = cc.RANDOM_0_1();
    return rnd*(max-min)+min;
};

CCShake.create = function(d, strength){
    return CCShake.createWithStrength( d, strength, strength );
};

CCShake.createWithStrength = function(duration, strength_x, strength_y){
    var pRet = new CCShake();
    if (pRet)
    {
        pRet.initWithDur(duration, strength_x, strength_y);
    }
    return pRet;
};